### PR TITLE
Timeouts

### PR DESCRIPTION
**Pull request by alexplischke**

### CODE REVIEW

File: monocle/stack/network/__init__.py [hunk 1]
```diff
@@ -21,25 +21,36 @@ def __init__(self, reader, writer):
...
 
     async def read_some(self) -> bytes:
         # there's no equivalent of this "read as much as can be read
         # quickly" in the asychio streaming API -- just set a
         # reasonable limit
-        return await self._reader.read(65536)
+        try:
+            return await asyncio.wait_for(self._reader.read(65536), self.timeout)
+        except asyncio.TimeoutError as e:
+            raise ConnectionLost() from e
```
Comment (sah) [inline monocle/stack/network/__init__.py]: How about making these `ConnectionLost("timeout")`?

File: monocle/stack/network/__init__.py [hunk 2]
```diff
@@ -21,25 +21,36 @@ def __init__(self, reader, writer):
...
-        self._closed_flag = False
+        self._closed = False
 
     async def read_some(self) -> bytes:
         # there's no equivalent of this "read as much as can be read
         # quickly" in the asychio streaming API -- just set a
         # reasonable limit
-        return await self._reader.read(65536)
+        try:
+            return await asyncio.wait_for(self._reader.read(65536), self.timeout)
```
Comment (sah) [inline monocle/stack/network/__init__.py]: Does this do the right thing when `self.timeout == None`?

File: monocle/stack/network/__init__.py (hunk 1 above)
Comment (sah) [reply monocle/stack/network/__init__.py]: Actually, second thought: is it important to wrap these exceptions? What would happen if we just let asyncio raise its own exception here? This could be something where forcing user code to deal with the new exceptions would gently encourage migration toward native APIs without causing too much initial work. But, I haven't researched what the impact would be or thought about this hard enough.

File: monocle/stack/network/__init__.py (hunk 2 above)
Comment (alexplischke) [reply monocle/stack/network/__init__.py]: Yup, then it's treated as "run without a timeout".

File: monocle/stack/network/__init__.py (hunk 1 above)
Comment (alexplischke) [reply monocle/stack/network/__init__.py]: I had similar thoughts and the effect would be exactly as you described. Essentially, all places where we catch ExceptionLost would have to be replaced by TimeoutError.
Honestly, likely not too crazy of a change, just a "search and replace" across whatever starts using tricle. Which honestly isn't in that many places. There might be some semantic nuance here and there (timeout vs. actual loss of connection).

File: monocle/stack/network/__init__.py [hunk 3]
```diff
@@ -11,6 +11,13 @@
 
 
 class ConnectionLost(Exception):
+    """ *** DEPRECATED ***
+        Please use ConnectionTimeout instead.
+    """
```
Comment (sah) [inline monocle/stack/network/__init__.py]: Wait, is this actually never used for anything except timeouts?

Comment (alexplischke) [reply monocle/stack/network/__init__.py]: Unfortunately, the code in sauce doesn't seem to make a clear distinction and sometimes assumes it's a timeout, in some other cases it doesn't care and treats it as "well, I couldn't connect or execute my command. let me try again".
From the technical perspective, `ConnectionLost` could actually mean both timeout and loss of connection. If our code timed out, it meant timeout. If twisted/tornado closed the connection, we wrapped it in ConnectionLost and issued a callback with the original reason attached. But yeah, our code didn't seem to care either way and handled it more like a generic connection exception.

Comment (sah) [reply monocle/stack/network/__init__.py]: Seems like this comment is not quite right, then, if there are cases where this really does mean the connection is lost and it's the correct exception in those cases. I do think the original intent here was to make this a generic "connection problem" exception.

Comment (alexplischke) [reply monocle/stack/network/__init__.py]: This is getting interesting :)
If we look at it from a different perspective, we know that python will be throwing us some kind of `ConnectionError`, like `ConnectionResetError` (one of the builtins) if the connection is closed when we try to write to it.
The only thing that we do on top is, as of right now, to time the request out, thus adding asyncio.TimeoutError to the mix.

Our options are seemingly either:
1. Pass those exceptions straight up to the caller.
2. Introduce a similar hierarchy of exceptions to the client

Point 1) is self explanatory.

For point 2) we could add something like this:
- ConnectionException (our generic exception)
- RequestTimeout (SocketTimeout, ConnectionTimeout, RequestTimeout, you name it :))
However, the question as what to do with `ConnectionLost` would then remain. Either we keep it around (and wrap python's ConnectionError with it) or we rename it to the generic ConnectionException.

The current code does not throw `ConnectionLost` under any circumstance. So if we want to keep it around, we'd have to choose to deal with it in some form as described in Point 2).

Comment (alexplischke) [reply monocle/stack/network/__init__.py]: I’m currently favoring this:
1. Rename `ConnectionLost` to `ConnectionError`
2. Add `ConnectionTimeout`
3. If timeout occurs, raise `ConnectionTimeout`
4. Any other exception is a straight up `ConnectionError`.

File: monocle/stack/network/__init__.py [hunk 4]
```diff
@@ -10,36 +10,55 @@
 from monocle.stack import eventloop
 
 
-class ConnectionLost(Exception):
+class ConnectionException(ConnectionError):
```
Comment (sah) [inline monocle/stack/network/__init__.py]: do you mean `asyncio.ConnectionError`?

Comment (alexplischke) [reply monocle/stack/network/__init__.py]: That doesn't exist to my knowledge :o Tried to produce an exception and I was getting the builtin `ConnectionResetError`, which inherits the builtin `ConnectionError`.

Looking into some asyncio packages, I see that `asyncio.streams.IncompleteReadError` can also be thrown as well (which inherits builtin `EOFError`).
